[Feature](cte) support recursive cte

BiteTheDDDDt opened the pull request:
### What problem does this PR solve?
https://github.com/apache/doris/issues/44550

This pull request introduces support for recursive Common Table Expressions (CTEs) in the pipeline execution engine. It adds several new operators and supporting code for handling recursive CTEs, as well as some refactoring and minor improvements to existing operator infrastructure.

Key changes include:

**Recursive CTE Operator Support:**

* Added new pipeline operators for recursive CTE handling: `RecCTEAnchorSinkOperatorX`, `RecCTESinkOperatorX`, `RecCTEScanOperatorX`, and `RecCTESourceOperatorX`, along with their corresponding local state classes and shared state (`RecCTESharedState`). These operators manage the anchor and recursive branches of a CTE, data exchange, and block dependencies. [[1]](diffhunk://#diff-a3c735f00b624acbb7121686c4a301aefd099f98d7f3a1bc70c47f8fc4ba7ad6R1-R56) [[2]](diffhunk://#diff-29beb8e5e497a40e705584cc48be29d9bfa77e9a784a667ffefd0d936fa80989R1-R128) [[3]](diffhunk://#diff-b9a9698872bce79149b840687e249161ccbdea3fdde4aeae5ed6f5c0ec2f937fR1-R55) [[4]](diffhunk://#diff-51f0fff6b8c555917b3e55fe935928ec4c845ffbfbe365841ca57f68dca971bfR1-R89)
* Registered the new recursive CTE operators and their local state classes in the operator factory and template instantiations to ensure they are recognized by the pipeline engine. [[1]](diffhunk://#diff-8922a03a948ba6a44dda51be21f535e0f5b8725b37b2f10facc90f2476b82aebR810-R811) [[2]](diffhunk://#diff-8922a03a948ba6a44dda51be21f535e0f5b8725b37b2f10facc90f2476b82aebR845-R846) [[3]](diffhunk://#diff-8922a03a948ba6a44dda51be21f535e0f5b8725b37b2f10facc90f2476b82aebR882) [[4]](diffhunk://#diff-8922a03a948ba6a44dda51be21f535e0f5b8725b37b2f10facc90f2476b82aebR900)
* Included the necessary header files for the new recursive CTE operators in the operator implementation file.

**Operator Infrastructure Improvements:**

* Added a `need_rerun` virtual method to `DataSinkOperatorXBase` to allow sink operators to specify whether they require rerunning, which can be used for iterative algorithms like recursive CTEs.

**Exchange Operator Enhancements & Bug Fixes:**

* Implemented a `reset` method for `ExchangeSourceOperatorX` and `ExchangeLocalState`, allowing the exchange source operator to be reset and reused, which is useful for iterative execution in recursive CTEs. [[1]](diffhunk://#diff-2074d2cbf58e5136dd6b0a7e1dbe13396aaa97efecfc366e2506e71bbbe7411aR51-R67) [[2]](diffhunk://#diff-2074d2cbf58e5136dd6b0a7e1dbe13396aaa97efecfc366e2506e71bbbe7411aR103-R104) [[3]](diffhunk://#diff-1e274a427d9138d0c5d6d9916fa9c4918d8be155e8bf070686a9def45c23a0d8R228-R232)
* Improved robustness in `ExchangeLocalState::debug_string` by handling the case where `stream_recvr` may be null, and ensured `stream_recvr` is reset after closing to avoid dangling pointers. [[1]](diffhunk://#diff-1e274a427d9138d0c5d6d9916fa9c4918d8be155e8bf070686a9def45c23a0d8L53-R53) [[2]](diffhunk://#diff-1e274a427d9138d0c5d6d9916fa9c4918d8be155e8bf070686a9def45c23a0d8R213)

**Minor Refactoring:**

* Updated constructors for `ExchangeSinkBuffer` to pass an explicit flag to `HasTaskExecutionCtx`, clarifying context initialization. [[1]](diffhunk://#diff-1d297fbd1be58030c0f1c0193211c3a295bfa553ad412dea00beb4757c34ecc4L95-R95) [[2]](diffhunk://#diff-a4592d017ead7e98a51753f78f9e80336f4ce6f24262e9c488182090bcff0877L272-R272)

These changes collectively enable the pipeline engine to support recursive CTE queries, improve operator reusability, and enhance code safety and maintainability.

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [x] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [x] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [x] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

Commit:
recursive cte be part

starocean999 committed:
recursive cte fe part

fix some bug in fe

fix fe plan

update fe code

update fe

update fe part

update fe

add ut

starocean999 committed:
remove useless case

Commit:
support rec cte (be part / proto part)

Thearas commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

BiteTheDDDDt commented:
run buildall

starocean999 committed:
rebase code

BiteTheDDDDt commented:
run buildall

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  71.51% (640/895)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/58916_7e6d2071dc15724d93e60fc8ba978ed51a25cb0f_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/58916_7e6d2071dc15724d93e60fc8ba978ed51a25cb0f_merge_fe/report/index.html)

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  71.51% (640/895)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/58916_7e6d2071dc15724d93e60fc8ba978ed51a25cb0f_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/58916_7e6d2071dc15724d93e60fc8ba978ed51a25cb0f_merge_fe/report/index.html)

Commit:
update some fix, todo and comment

fix

fix

fix

Revert "fix"

This reverts commit 41bba83e370c028303bf77f808f6ab9908c6e0bd.

fix

update some todo and comment

update fix

update fix

test

tets

tets

remove nullable_tuples

update

format

update

update

fix

update

fix

fix

fix

Revert "fix"

This reverts commit ce862426090f4a2074be3e84a1815000ff3abc7a.

Reapply "fix"

This reverts commit 5c58d3d53bbebbe63e7660d7b22597df6242be31.

fix

update

fix

BiteTheDDDDt commented:
run buildall

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `91.43% (779/852)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/58916_eff7c8daf6a33f1376af1136057b41122b7db72d_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/58916_eff7c8daf6a33f1376af1136057b41122b7db72d_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 72.36% (25056/34628) |
| Line Coverage     | 59.13% (262985/444757) |
| Region Coverage   | 53.95% (218092/404248) |
| Branch Coverage   | 55.57% (93424/168117) |

BiteTheDDDDt commented:
run p0 5

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  70.73% (633/895)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/58916_eff7c8daf6a33f1376af1136057b41122b7db72d_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/58916_eff7c8daf6a33f1376af1136057b41122b7db72d_merge_fe/report/index.html)

starocean999 committed:
fix create mtmv task fail

starocean999 committed:
fix bug

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/6ff0e83295f42a45ce3891a2005d2c25e7f7fc4e_6ff0e83295f42a45ce3891a2005d2c25e7f7fc4e_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/6ff0e83295f42a45ce3891a2005d2c25e7f7fc4e_6ff0e83295f42a45ce3891a2005d2c25e7f7fc4e_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.48% (1766/2222) |
| Line Coverage     | 64.79% (31234/48205) |
| Region Coverage   | 65.32% (15546/23798) |
| Branch Coverage   | 56.01% (8270/14764) |

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  66.30% (596/899)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/58916_6ff0e83295f42a45ce3891a2005d2c25e7f7fc4e/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/58916_6ff0e83295f42a45ce3891a2005d2c25e7f7fc4e/fe_report/index.html)

BiteTheDDDDt commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `9.21% (84/912)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/e40279bc930d1bf3d021450cbee09444c44288a3_e40279bc930d1bf3d021450cbee09444c44288a3/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/e40279bc930d1bf3d021450cbee09444c44288a3_e40279bc930d1bf3d021450cbee09444c44288a3/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 53.24% (18944/35582) |
| Line Coverage     | 39.15% (175570/448428) |
| Region Coverage   | 33.68% (135597/402623) |
| Branch Coverage   | 34.63% (58594/169224) |

zfr9527 committed:
[test](cte)add recursive cte test case (#59352)

[test](cte)add recursive cte test case

### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
- [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
- [ ] Yes. <!-- Add document PR link here. eg:
https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR
should merge into -->

zfr9527 commented:
run buildall

zfr9527 committed:
fix recursive cte p0 case (#59402)

fix recursive cte p0 case

[test](cte)add recursive cte test case

### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

### Release note

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
- [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
- [ ] Yes. <!-- Add document PR link here. eg:
https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR
should merge into -->

BiteTheDDDDt commented:
run buildall

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  66.30% (596/899)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/58916_1ef50682c751f43f2c23e42ff0e6ed290a95ad61/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/58916_1ef50682c751f43f2c23e42ff0e6ed290a95ad61/fe_report/index.html)

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  75.64% (680/899)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/58916_1ef50682c751f43f2c23e42ff0e6ed290a95ad61_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/58916_1ef50682c751f43f2c23e42ff0e6ed290a95ad61_merge_fe/report/index.html)

HappenLee reviewed (dismissed):
LGTM

Commit:
Merge branch 'master' into dev_rec4

BiteTheDDDDt commented:
run buildall

Commit:
fix thrift conflict

BiteTheDDDDt commented:
run buildall

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRecursiveCte.java:
```diff
@@ -0,0 +1,284 @@
+    private final String cteName;
+    private final List<NamedExpression> outputs;
+    private final List<List<SlotReference>> regularChildrenOutputs;
+    private final boolean isUnionAll;
```
isUnionAll -> reuse Qualifier in set operation

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRecursiveCteScan.java:
```diff
@@ -0,0 +1,87 @@
+/**
+ * LogicalRecursiveCteScan.
+ */
+public class LogicalRecursiveCteScan extends LogicalCatalogRelation {
```
在pg中称为 WorkTableReference

HappenLee reviewed (approved):
LGTM

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/glue/translator/PhysicalPlanTranslator.java:
```diff
@@ -1037,6 +1042,28 @@ public PlanFragment visitPhysicalSchemaScan(PhysicalSchemaScan schemaScan, PlanT
+
+        translateRuntimeFilter(recursiveCteScan, scanNode, context);
+
+        context.addScanNode(scanNode, recursiveCteScan);
```
i don't think it should be add to scan node. it is not event a real scan node

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:
```diff
@@ -1127,6 +1127,8 @@ public class LogicalPlanBuilder extends DorisParserBaseVisitor<Object> {
 
     private final Map<Integer, ParserRuleContext> selectHintMap;
 
+    private boolean isInRecursiveCteContext = false;
```
add comment to explain why need this flag

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/properties/ChildOutputPropertyDeriver.java:
```diff
@@ -150,6 +153,11 @@ public PhysicalProperties visitPhysicalFileScan(PhysicalFileScan fileScan, PlanC
 
+    @Override
+    public PhysicalProperties visitPhysicalRecursiveCteScan(PhysicalRecursiveCteScan cteScan, PlanContext context) {
+        return PhysicalProperties.ANY;
```
if this is any, we should not do compute scan range for CteScanNode when translator, and let coordinator choose a random BE

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/AnalyzeCTE.java:
```diff
@@ -95,25 +109,146 @@ private Pair<CTEContext, List<LogicalCTEProducer<Plan>>> analyzeCte(
+            CTEContext outerCteCtx, CascadesContext cascadesContext) {
+        Preconditions.checkArgument(aliasQuery.isRecursiveCte(), "alias query must be recursive cte");
+        LogicalPlan parsedCtePlan = (LogicalPlan) aliasQuery.child();
+        if (!(parsedCtePlan instanceof LogicalUnion) || parsedCtePlan.children().size() != 2) {
```
```suggestion
        if (!(parsedCtePlan instanceof LogicalUnion) || parsedCtePlan.arity() != 2) {
```

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/AnalyzeCTE.java:
```diff
@@ -95,25 +109,146 @@ private Pair<CTEContext, List<LogicalCTEProducer<Plan>>> analyzeCte(
+        // analyze anchor child, its output list will be recursive cte temp table's schema
+        LogicalPlan anchorChild = (LogicalPlan) parsedCtePlan.child(0);
+        CascadesContext innerAnchorCascadesCtx = CascadesContext.newContextWithCteContext(
+                cascadesContext, anchorChild, outerCteCtx, Optional.of(aliasQuery.getAlias()), ImmutableList.of());
```
why need cte name when analyze anchor side?

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/AnalyzeCTE.java:
```diff
@@ -95,25 +109,146 @@ private Pair<CTEContext, List<LogicalCTEProducer<Plan>>> analyzeCte(
+        if (recursiveCteScanList.size() > 1) {
+            throw new AnalysisException(String.format("recursive reference to query %s must not appear more than once",
+                    aliasQuery.getAlias()));
+        }
```
we do not support nested recursive cte in recursive side?

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/AnalyzeCTE.java:
```diff
@@ -95,25 +109,146 @@ private Pair<CTEContext, List<LogicalCTEProducer<Plan>>> analyzeCte(
+        }
+        List<Slot> recursiveChildOutputs = analyzedRecursiveChild.getOutput();
+        for (int i = 0; i < recursiveChildOutputs.size(); ++i) {
+            if (!recursiveChildOutputs.get(i).getDataType().equals(anchorChildOutputTypes.get(i))) {
```
think about two struct type, that have same field number and field type, only have different field name. another case is variant type, check variant type's equals function, it has some attributes, we should allow these attributes different because struct's fields' name and variant's attribute are may be generated by planner itself and could be changed between different version

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/AnalyzeCTE.java:
```diff
@@ -95,25 +109,146 @@ private Pair<CTEContext, List<LogicalCTEProducer<Plan>>> analyzeCte(
+            }
+        }
+        analyzedRecursiveChild = new LogicalRecursiveCteRecursiveChild<>(aliasQuery.getAlias(),
+                forceOutputNullable(analyzedRecursiveChild, ImmutableList.of()));
```
why also changed recursive side's output name?

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/AnalyzeCTE.java:
```diff
@@ -95,25 +109,146 @@ private Pair<CTEContext, List<LogicalCTEProducer<Plan>>> analyzeCte(
+        LogicalRecursiveCte analyzedCtePlan = new LogicalRecursiveCte(aliasQuery.getAlias(),
+                logicalUnion.getQualifier() == SetOperation.Qualifier.ALL,
+                ImmutableList.of(analyzedAnchorChild, analyzedRecursiveChild));
+        List<List<NamedExpression>> childrenProjections = analyzedCtePlan.collectChildrenProjections();
```
i don't think recursive cte need this step, because it used to do cast before set operation, but all output of children of recursive cte are excactly same, so we do not need to do cast.

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java:
```diff
@@ -171,16 +173,32 @@ private LogicalPlan bindWithCurrentDb(CascadesContext cascadesContext, UnboundRe
-        TableIf table = cascadesContext.getStatementContext().getAndCacheTable(tableQualifier, TableFrom.QUERY,
-                Optional.of(unboundRelation));
+        LogicalPlan scan;
+        if (tableName.equalsIgnoreCase(cascadesContext.getCurrentRecursiveCteName().orElse(""))) {
```
should not use equalsIgnoreCase here, lower_case_table_names also effect cte binding, so we should use `org.apache.doris.nereids.CTEContext#findCTEContext` here. why not reuse normal cte binding code?

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/AdjustNullable.java:
```diff
@@ -289,6 +290,53 @@ public Plan visitLogicalRepeat(LogicalRepeat<? extends Plan> repeat, Map<ExprId,
+        if (!recursiveCte.children().isEmpty()) {
+            inputNullable = Lists.newArrayListWithCapacity(recursiveCte.getOutputs().size());
+            for (int i = 0; i < recursiveCte.getOutputs().size(); i++) {
+                inputNullable.add(false);
```
these code block let me confused, in which scenario, LogicalRecursiveCte' children's output could be not null? or it is just defensive programming

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ColumnPruning.java:
```diff
@@ -403,6 +410,54 @@ public <P extends Plan> P pruneOutput(P plan, List<NamedExpression> originOutput
         }
     }
 
+    private LogicalRecursiveCte pruneRecursiveCteOutput(LogicalRecursiveCte recursiveCte, PruneContext context) {
```
remove this function, it useless

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/CTEInline.java:
```diff
@@ -113,4 +128,19 @@ public Plan visitLogicalCTEConsumer(LogicalCTEConsumer cteConsumer, LogicalCTEPr
+                cteConsumers.add((LogicalCTEConsumer) planNode);
+            }
+        } else if (planNode instanceof LogicalRecursiveCteRecursiveChild) {
+            collectMustInlineCteConsumers(planNode.child(0), true, cteConsumers);
```
why need this collector? i think we could add a recursive flag on CteAnchor or somewhere else to process it easily

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalRecursiveCteRecursiveChild.java:
```diff
@@ -0,0 +1,117 @@
+/**
+ * LogicalRecursiveCteRecursiveChild is sentinel plan for must_shuffle
+ */
+public class LogicalRecursiveCteRecursiveChild<CHILD_TYPE extends Plan> extends LogicalUnary<CHILD_TYPE> {
```
should implement hashCode and equals

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRecursiveCte.java:
```diff
@@ -0,0 +1,292 @@
+    }
+
+    @Override
+    public List<Slot> computeOutput() {
```
physical node should not impl computeOutput

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalRecursiveCteRecursiveChild.java:
```diff
@@ -0,0 +1,121 @@
+
+    @Override
+    public void computeEqualSet(DataTrait.Builder builder) {
+
```
i'm confused if this node do not impl any data traits functions, why PhysicalRecursiveCte need to implement these functions?

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/qe/runtime/ThriftPlansBuilder.java:
```diff
@@ -576,6 +595,130 @@ private static void filterInstancesWhichReceiveDataFromRemote(
         }
     }
 
+    private static Set<Integer> setParamsForRecursiveCteNode(List<PipelineDistributedPlan> distributedPlans,
```
this function is too long, please add some comment to it to explain how it is work
